### PR TITLE
Interpolate

### PR DESCRIPTION
**Pull request by oxinabox**
This shows how interpolation could be done.

I don't now if it meets the API spec right now.
But I think has the ML side covered,
so I am hoping I can leave it to @ChrisRackauckas 
to wrap up the DiffEq stuff, 
and to know how to actually tests it.

BTW this now works with **malmaud/TensorFlow.jl:master**

### CODE REVIEW

File: src/TensorFlowDiffEq.jl [hunk 1]
```diff
@@ -14,6 +14,46 @@ odetf(;hl_width=64) = odetf(hl_width)
...
+    as_default(interp.sess.graph) do
+        prev = grad_node(interp, Val{N-1})
+        tt = interp.sess.graph["tt"] #load the `tt` node from the graph
+        grads(prev, tt)
+    end
+end
+
+function (id::TensorFlowInterpolation){N}(tvals, idxs, deriv::Type{Val{N}})
+    gn = grad_node(id, derviv)
+    vals = run(id.sess, gn, Dict(t=>tvals))
```
Comment (ChrisRackauckas) [inline src/TensorFlowDiffEq.jl:38]: does this work with `tvals <: Union{Number,AbstractArray}`?

File: src/TensorFlowDiffEq.jl [hunk 2]
```diff
@@ -14,6 +14,46 @@ odetf(;hl_width=64) = odetf(hl_width)
...
+    #PREM-OPT: the indexing could be moved inside the network, to avoid even calculating gradients for columns are are not using, by first slicing `interp.u` inside the `grad_node` function.
+    vals[:,idxs]'
+end
+
+function (id::TensorFlowInterpolation){N}(v, tvals, idxs, deriv::Type{Val{N}})
+    # In-place version, noting that truely inplace operations between julia and tensorflow are actually impossible
```
Comment (ChrisRackauckas) [inline src/TensorFlowDiffEq.jl]: Sad to hear.

**Comment by ChrisRackauckas** [comment]
Just need to make sure that it works with `tvals` a Number as well as an AbstractArray and I'll merge. Otherwise it just needs a line to wrap it in an array.

**Comment by oxinabox** [comment]
I'm pretty sure it would not work on a scalar. So I added that line.

`t` is expected to be a vector,
which is expanded into a matrix `tt`.

It might work on a scalar once it is rewritten to all be transposed.

I am interested to see the results of testing this

File: src/TensorFlowDiffEq.jl [hunk 3]
```diff
@@ -14,6 +14,48 @@ odetf(;hl_width=64) = odetf(hl_width)
...
+end
+
+function (id::TensorFlowInterpolation){N}(tvals, idxs, deriv::Type{Val{N}})
+    gn = grad_node(id, derviv)
+    vals = run(id.sess, gn, Dict(t=>tvals))
+    #PREM-OPT: the indexing could be moved inside the network, to avoid even calculating gradients for columns are are not using, by first slicing `interp.u` inside the `grad_node` function.
+    vals[:,idxs]'
+end
+
+(id::TensorFlowInterpolation)(tval::Number, idxs, deriv) = id([tval], idxs, deriv)
```
Comment (ChrisRackauckas) [inline src/TensorFlowDiffEq.jl:43]: inplace version too?

File: src/TensorFlowDiffEq.jl (hunk 2 above)
Comment (oxinabox) [reply src/TensorFlowDiffEq.jl]: I can't fully rule-out being able to define `run!(outs_jl, outs_tfnode, ...)`
I'll double check that.

In general, copying has to be done to put things onto and take off of the GPU anyway.

Comment (ChrisRackauckas) [reply src/TensorFlowDiffEq.jl]: >In general, copying has to be done to put things onto and take off of the GPU anyway.

That makes sense.

Comment (ChrisRackauckas) [reply src/TensorFlowDiffEq.jl]: Though couldn't you `copy!` into a user defined array to bring things back from the GPU?

File: src/TensorFlowDiffEq.jl (hunk 3 above)
Comment (oxinabox) [reply src/TensorFlowDiffEq.jl:43]: Yeah that should be fine already,
 the 4 arg version (inplace) always calls one of the two  3 arg version (`tvals::Number` or `tvals::Any`).

Comment (ChrisRackauckas) [reply src/TensorFlowDiffEq.jl:43]: oh I missed that.

File: src/TensorFlowDiffEq.jl (hunk 2 above)
Comment (oxinabox) [reply src/TensorFlowDiffEq.jl]: true, (in theory).
Which would bring 2 copies down to 1.

Comment (ChrisRackauckas) [reply src/TensorFlowDiffEq.jl]: still worthwhile.

Comment (oxinabox) [reply src/TensorFlowDiffEq.jl]: yeah, looking at it it might be possible.

**Pull request merged**